### ISSUE TITLE
Can't allocate 'shared' memory for msg for channel id

### ISSUE THREAD

**Issue by lkuty** [slact/nchan#147]
Hi,

More or less every day, nchan stops working properly with errors related to the allocation of memory. I use version 0.99.8 and nginx 1.8.1. A few messages are sent every 30sec (less than 10). I use the default shared memory size of 32768KiB.

The error log for nginx contains lines like those ones :

    2016/03/24 11:41:36 [crit] 19805#0: ngx_slab_alloc() failed: no memory
    2016/03/24 11:41:36 [error] 19805#0: shpool alloc failed
    2016/03/24 11:41:36 [error] 19805#0: MEMSTORE:01: can't allocate 'shared' memory for msg for channel id
    2016/03/24 11:41:36 [error] 19805#0: *18962 nchan: error publishing message, client: 192.168.0.13, server: _, request: "POST /pub/last/hun      HTTP/1.1", host: "192.168.0.10:8081"
    2016/03/24 11:41:36 [error] 19805#0: MEMSTORE:01: can't create shared message for channel /last_hun
    2016/03/24 11:41:36 [crit] 19805#0: ngx_slab_alloc() failed: no memory
    2016/03/24 11:41:36 [error] 19805#0: shpool alloc failed
    2016/03/24 11:41:36 [error] 19805#0: MEMSTORE:01: can't allocate 'shared' memory for msg for channel id
    2016/03/24 11:41:36 [error] 19805#0: *18963 nchan: error publishing message, client: 192.168.0.13, server: _, request: "POST /pub/energie/hun      HTTP/1.1", host: "192.168.0.10:8081"
    2016/03/24 11:41:36 [error] 19805#0: MEMSTORE:01: can't create shared message for channel /energie_hun
    2016/03/24 11:41:36 [crit] 19805#0: ngx_slab_alloc() failed: no memory
    2016/03/24 11:41:36 [error] 19805#0: shpool alloc failed
    2016/03/24 11:41:36 [crit] 19805#0: ngx_slab_alloc() failed: no memory
    2016/03/24 11:41:36 [error] 19805#0: shpool alloc failed
    2016/03/24 11:41:36 [error] 19805#0: MEMSTORE:01: can't allocate 'shared' memory for msg for channel id
    2016/03/24 11:41:36 [alert] 21080#0: worker process 19805 exited on signal 11 (core dumped)

I use SSE. I have three publishers with a config that looks like this : 

    location ~ /pub/last/(\w+)$ {
         nchan_publisher http;
         nchan_channel_id last_$1;
         nchan_message_buffer_length 1;
         nchan_message_timeout 0;
     }

And for the corresponding subscriber, the config looks like this :

    location ~ /sub/last/(\w+)$ {
      nchan_subscriber eventsource;
      nchan_channel_id last_$1;
    }

Maybe a memory leak somewhere ?

**Comment by slact** [comment]
Hmm, it might have to do with `nchan_message_timeout 0;`
As an experiment, could you set it to something large, like a couple of days, and see if this issue persists?
Also, how large are the messages you're publishing? What does the rest of your nginx config look like? What OS are you using? What CPU?

**Comment by lkuty** [comment]
I've put 3d for `nchan_message_timeout`. The messages are less than 3200 bytes. The OS is Ubuntu 14.04.3 LTS x64. The CPU is Quad-Core AMD Opteron Processor 1385.

The nginx config for the publishers is :

    server {
      server_name _;
      server_tokens off;
    
      listen 0.0.0.0:8081;
    
      root /home/hydro/web/public;
    
      proxy_buffering off;
      proxy_cache off;
    
      proxy_pass_request_headers on;
      proxy_set_header X-Real-IP  $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;
      proxy_set_header X-Remote-User $remote_user;
      proxy_redirect off;
    
      location ~ /pub/last/(\w+)$ {
          nchan_publisher http;
          nchan_channel_id last_$1;
          nchan_message_buffer_length 1;
          nchan_message_timeout 3d;
      }
    
      location ~ /pub/energie/(\w+)$ {
          nchan_publisher http;
          nchan_channel_id energie_$1;
          nchan_message_buffer_length 1;
          nchan_message_timeout 3d;
      }
    
      location ~ /pub/cmd/(\w+)$ {
          nchan_publisher http;
          nchan_channel_id cmd_$1;
          nchan_message_buffer_length 1;
          nchan_message_timeout 3d;
      }
    }

The configuration for the web app is :

    server {
      if (-f $document_root/in_maintenance_mode) {
        return 503;
      }
    
      server_name _;
      server_tokens off;
    
      listen 0.0.0.0:443 default ssl;
    
      root /home/hydro/web/public;
    
      location / {
        proxy_buffering off;
        proxy_cache off;
    
        proxy_pass_request_headers on;
        passenger_set_header Host $http_host;
        passenger_set_header X-Real-IP  $remote_addr;
        passenger_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        passenger_set_header X-Forwarded-Proto $scheme;
        passenger_set_header X-Remote-User $remote_user;
        passenger_set_header Host $http_host;
    
        passenger_min_instances 3;
    
        proxy_redirect off;
        passenger_enabled on;
        passenger_ruby /home/hydro/.rbenv/versions/2.3.0/bin/ruby;
        passenger_load_shell_envvars on;
        passenger_nodejs /usr/bin/nodejs;
        passenger_friendly_error_pages on;
        passenger_app_env development;
      }
    
      location ~* \.(js|css|png|gif|jpg|jpeg|txt|pdf|zip|rar|iso|7z|html|ico|map)$ {
      }
    
      location ~ /sub/last/(\w+)$ {
        nchan_subscriber eventsource;
        #nchan_channel_id all_$remote_user;
        nchan_channel_id last_$1;
      }
    
      location ~ /sub/energie/(\w+)$ {
        nchan_subscriber eventsource;
        #nchan_channel_id all_$remote_user;
        nchan_channel_id energie_$1;
      }
    
      location ~ /sub/cmd/(\w+)$ {
        nchan_subscriber eventsource;
        #nchan_channel_id all_$remote_user;
        nchan_channel_id cmd_$1;
      }
    
      error_page 503 @maintenance;
    
      location @maintenance {
        rewrite ^(.*)$ /503.html break;
      }
    }

The main nginx config is :

    user www-data;
    worker_processes 4;
    pid /run/nginx.pid;
    env PATH;
    
    events {
      worker_connections 768;
    }
    
    http {
      sendfile on;
      tcp_nopush on;
      tcp_nodelay on;
      keepalive_timeout 65;
      types_hash_max_size 2048;
    
      include /opt/nginx/conf/mime.types;
      default_type application/octet-stream;
    
      ssl_protocols TLSv1 TLSv1.1 TLSv1.2; # Dropping SSLv3, ref: POODLE
      ssl_prefer_server_ciphers on;
    
      access_log /opt/nginx/logs/access.log;
      error_log /opt/nginx/logs/error.log;
    
      gzip on;
      gzip_disable "msie6";
    
      passenger_root /opt/nginx_modules/passenger-5.0.26;
      passenger_ruby /home/hydro/.rbenv/versions/2.3.0/bin/ruby;
      passenger_log_file /opt/nginx/logs/passenger.log;
      passenger_max_pool_size 6;
    
      include /opt/nginx/conf/conf.d/*.conf;
      include /opt/nginx/conf/sites-enabled/*;
    }

I have removed any configuration related to SSL or HTTP Basic Auth. And also public server names.

**Comment by slact** [comment]
Thank you. Let me know what happens tomorrow.

**Comment by lkuty** [comment]
So far, everything is ok. I am considering setting `nchan_store_messages` to `off` which IMHO means that messages will be sent to current subscribers and not buffered for new subscribers. Not a problem in my case because messages are sent regularly and the client first make an Ajax call to get data (without using SSE). I could also set `nchan_message_timeout` to a short time like `1m`.

**Comment by slact** [comment]
`nchan_store_messages off` means that if you're using long-polling or interval-polling, you'll probably be losing messages. That is the nature of the protocol -- the subscriber is enqeued for only one response, and if a message is published while a subscriber is sending a new request, it will be lost. So I don't recommend it.

I haven't yet looked at the code regarding the bug you've found, but I suspect it's a pretty simple fix. `nchan_message_timeout 0;` just needs some special-case handling. I will update when this is resolved, which will be in a few days. (I'm currently busy refactoring a lot of the subscriber output code for another bugfix (#132)

**Comment by slact** [comment]
Fixed in v0.99.10

**Comment by slact** [comment]
I'm closing this issue, please reopen if you're still having problems.

**Comment by lkuty** [comment]
OK, thanks for your fix ! :) Eventually, the problem arose again after a few days so I set `nchan_store_messages off` and `nchan_message_timeout 5m`. I will let those parameters like that to see if the problem apperas again before upgrading to v0.99.10 and maybe reverting to the old parameters.

**Comment by slact** [comment]
the issue actually affected all timeout values -- messages weren't freed from memory until the timeout even after they were deleted from the channel. So this should have fixed the `5m` case as well. Try it out and let me know what you find. Thanks for your feedback!

**Comment by lpenguin** [comment]
Hi @slact ! I have memory leak problem too. Here is my configuration:
Ubuntu 14.04.4 LTS, 4GB Dual Core AWS  Medium Instance
NChan Version: 0.99.16
nginx version: 1.11.3

nginx config file:

```
server {
    listen 80;

    
    nchan_store_messages off;
    nchan_message_timeout 0;
    nchan_subscriber_first_message newest;


    location ~ /rt/clients/(\w+)$ {
        nchan_pubsub;
        nchan_publisher_channel_id sr$1;
        nchan_subscriber_channel_id ra$1;
    }

    location ~ /rt/server/(\w+)$ {
        nchan_pubsub;
        nchan_publisher_channel_id ra$1;
        nchan_subscriber_channel_id sr$1;
    }
    location ~ /rt/server/pub/(\w+)$ {
        nchan_publisher;
        nchan_channel_id ra$1;
    }
    location ~ /rt/server/sub/(\w+)$ {
        nchan_subscriber;
        nchan_channel_id sr$1;
    }

}
```
Currently i work only with  `~ /rt/clients/(\w+)` location.
Tried to change with `nchan_message_timeout`, but it does not help..

**Comment by slact** [comment]
How large are your channel ids, and how many channels are there? Shared memory is used for messages as well as some channel metadata and the id strings.

Also, what error are you getting?

**Comment by lpenguin** [comment]
Currently I use 1 channel

Here is my error.log
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 56
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 56
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100
2016/07/29 14:41:38 [error] 30633#0: MEMSTORE:01: force-reaping msg with refcount 100

**Comment by slact** [comment]
can you show me your nginx config? Edit out whatever you need for privacy. How large are the messages, and how many get published per second?

**Comment by lpenguin** [comment]
message is a json string about 100bytes. i publish about 30 messages per second

**Comment by slact** [comment]
I can't reproduce this so far, so either I need to see this happening on your machine, or you need to describe what publishers and subscribers you have at what location, how many of them there are, what kind, and what they are doing in as much detail as possible.

**Comment by lpenguin** [comment]
Here is my [nginx config](https://gist.github.com/lpenguin/5abcb42ff77afcf3ac4050f7edf22f02#file-nginx-conf)
I use only one location - `~ /rt/clients/(\w+)$`
I have 100 bots, each of them publish into this location one message each 1.5-2.5 seconds
So it's ~30 msg per second
Each message about 100 characters long
Here is my script that runs bots - [reaction_bots.go](https://gist.github.com/lpenguin/5abcb42ff77afcf3ac4050f7edf22f02#file-reaction_bots-go)

Please follow [this instruction](https://gist.github.com/lpenguin/5abcb42ff77afcf3ac4050f7edf22f02#file-install-md) to reproduce my issue. 

Also i made an Amazon Image [(ami-5c30fe3c)](https://console.aws.amazon.com/ec2/v2/home?region=us-west-2#LaunchInstanceWizard:ami=ami-5c30fe3c) with my configuration, just follow [run nginx and run go bots steps](https://gist.github.com/lpenguin/5abcb42ff77afcf3ac4050f7edf22f02#run-nginx). t2-nano instance type will be enough, it's cheapest one.

**Comment by slact** [comment]
Thanks for the detailed instructions, running your tests now.

**Comment by slact** [comment]
Found it, fixed in c8c8652fef3794fa . Rebuild from master and give it another try. This was a leak triggered by publishing via a websocket pubsub location. Thanks for your very detailed instructions. Top notch bug report!

**Comment by lpenguin** [comment]
Hi, slact! Thanks you, but unfornunately i keep getting "can't allocate 'shared' memory for msg" messages.. 

```
2016/08/04 21:02:31 [error] 9#0: shpool alloc failed
2016/08/04 21:02:31 [error] 9#0: MEMSTORE:00: can't allocate 'shared' memory for msg for channel id
2016/08/04 21:02:31 [error] 9#0: MEMSTORE:00: can't create unbuffered message for channel /srbots
2016/08/04 21:02:31 [crit] 9#0: ngx_slab_alloc() failed: no memory
```

I tested on different configuration: 100 clients, client send message every 0.05 sec. so it `1/0.05*100 = 2000` messages per second. I slightly changed client code, please use updated [install instructions](https://gist.github.com/lpenguin/5abcb42ff77afcf3ac4050f7edf22f02#file-install-md) and [nginx config](https://gist.github.com/lpenguin/5abcb42ff77afcf3ac4050f7edf22f02#file-nginx-conf)

**Comment by slact** [comment]
That's a different error. Will investigate.

**Comment by lpenguin** [comment]
Ok, thanks! By the way, i am running tests on [c8c8652 commit](https://github.com/slact/nchan/commit/c8c8652fef3794fa6d1790fdef595103b69d93ce), not on master

**Comment by slact** [comment]
This isn't a bug so much as a misleading config setting. `nchan_store_messages off` doesn't actually disable message storage, it just sets each channel's buffer length to 0. The messages themselves are still temporarily stored (for 5 seconds) in shared memory, separate from the channels. 

Here's what the shared memory block looks like when running tests with 32MB of reserved memory: 
https://asciinema.org/a/5sirr7c8t9vmfia6u5iov73tq
it fills up, then when the test is aborted, it gets freed up.

Here's the same test with 128MB:
https://asciinema.org/a/bwcvgtbhxzfvnopydj2g04gqs
Running smoothly this time.

**Comment by lpenguin** [comment]
ok, how can i change size of reserved memory (to 128MB)?

**Comment by slact** [comment]
[nchan_max_reserved_memory](https://nchan.slact.net/#nchan_max_reserved_memory) 128M;

**Comment by lpenguin** [comment]
thanks! no more error messages now in log
i had to look at the nchan docs first 😄 

so, if i want to handle more messages per sec - i have to increase nchan_max_reserved_memory?

**Comment by slact** [comment]
Yep.

**Comment by slact** [comment]
@lpenguin : I dug around a little more and it turns out I had misconfigured the unbuffered message garbage collector to be far lazier than it should have been. (fixed in 50c6374 )

Here is your test running in 32MB of shared memory after the fix: https://asciinema.org/a/f0p1auvx1i1lwbl11ej4x20c5

**Comment by solofeed** [comment]
@slact have similar error
![image](https://user-images.githubusercontent.com/10532978/58239510-dae87080-7d51-11e9-8abf-5cdfd9e0e54d.png)
Sometimes we got these errors, but we use Redis storage, and as I understand 
nchan_max_reserved_memory directive for default storage mode, by the way, we increased to 512bm, and it did not work

I suppose it might be the wrong configuration for nginx?
Stackoverflow says, that need to set up proper proxy cache setting
